File: src/Winium.Elements.Desktop/DesktopElement.cs
namespace Winium.Elements.Desktop
{
    #region using

    using System;
    using System.Collections.Generic;

    using OpenQA.Selenium;
    using OpenQA.Selenium.Remote;

    #endregion

    public abstract class DesktopElement : RemoteWebElement
    {
        #region Constructors and Destructors

        protected DesktopElement(IWebElement element)
            : base(GetRemoteWebDriver(element), element.GetId())
        {
        }

        #endregion

        #region Methods

        protected RemoteWebElement CreateRemoteWebElementFromResponse(Response response)
        {
            var elementDictionary = response.Value as Dictionary<string, object>;
            if (elementDictionary == null)
            {
                return null;
            }

            return new RemoteWebElement((RemoteWebDriver)this.WrappedDriver, (string)elementDictionary["ELEMENT"]);
        }

        private static RemoteWebDriver GetRemoteWebDriver(IWebElement element)
        {
            var remoteWebElement = element as RemoteWebElement;
            if (remoteWebElement == null)
            {
                throw new InvalidCastException("Specified cast is not valid. Please use RemoteWebElement as parameter.");
            }

            return (RemoteWebDriver)remoteWebElement.WrappedDriver;
        }

        #endregion
    }
}

File: java/src/main/java/winium/elements/desktop/DesktopElement.java
package winium.elements.desktop;

import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.remote.RemoteWebElement;
import org.openqa.selenium.remote.Response;
import winium.elements.desktop.extensions.WebElementExtensions;

import java.util.Map;

public abstract class DesktopElement extends RemoteWebElement {
    private static RemoteWebDriver getRemoteWebDriver(WebElement element) {
        if (!(element instanceof RemoteWebElement))
            throw new ClassCastException("Specified cast is not valid. Please use RemoteWebElement as parameter.");
        RemoteWebElement remoteWebElement = (RemoteWebElement)element;
        return (RemoteWebDriver)remoteWebElement.getWrappedDriver();
    }

    protected DesktopElement(WebElement element) {
        this.setParent(getRemoteWebDriver(element));
        this.setId(WebElementExtensions.getId(element));
    }

    protected RemoteWebElement createRemoteWebElementFromResponse(Response response) {
        Object value = response.getValue();
        if (!(value instanceof Map<?, ?>)) {
            return null;
        }
        Map<String, Object> elementDictionary = (Map<String, Object>)value;
        RemoteWebElement result = new RemoteWebElement();
        result.setParent((RemoteWebDriver)this.getWrappedDriver());
        result.setId((String)elementDictionary.get("ELEMENT"));
        return result;
    }
}
